File: leetcode/problems/cpp/sum-of-root-to-leaf-binary-numbers.cpp
// by @codeAbinash
// Time : O(n)
// Space : O(n)

struct TreeNode {
   int val;
   TreeNode* left;
   TreeNode* right;
   TreeNode() : val(0), left(nullptr), right(nullptr) {}
   TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
   TreeNode(int x, TreeNode* left, TreeNode* right) : val(x), left(left), right(right) {}
};

class Solution {
public:
   int sumRootToLeaf(TreeNode* root) {
      return dfs(root, 0);
   }
   int dfs(TreeNode* root, int sum) {
      if (!root) return 0;
      sum = sum << 1 | root->val;
      if (!root->left && !root->right) return sum;
      return dfs(root->left, sum) + dfs(root->right, sum);
   }
};

File: leetcode/problems/ts/sum-of-root-to-leaf-binary-numbers.ts
// by @codeAbinash
// Time : O(n)
// Space : O(n)

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function sumRootToLeaf(root: TreeNode | null): number {
  if (!root) return 0;
  let sum = 0;
  const dfs = (root: TreeNode, prev: number) => {
    const curr = (prev << 1) + root.val;
    if (!root.left && !root.right) {
      sum += curr;
      return;
    }
    root?.left && dfs(root.left, curr);
    root?.right && dfs(root.right, curr);
  };
  dfs(root, 0);
  return sum;
}
